pkg/base1/test-import-json.ts

```diff
@@ -0,0 +1,264 @@
+        [
+            'JSON validation error for : Not a boolean: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_array", function(assert) {
+    assert.deepEqual(import_array(["a", "b", "c"], import_string), ["a", "b", "c"], "array of strings");
+    assert.deepEqual(import_array([1, 2, 3], import_number), [1, 2, 3], "array of numbers");
+    assert.deepEqual(import_array([1, 2, "c"], import_number), [1, 2], "array of numbers with a string");
+    assert.throws(() => import_array("foo", import_string), "not an array");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for [2]: Not a number: "c"',
+            'JSON validation error for : Not an array: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_record", function(assert) {
+    assert.deepEqual(
+        import_record({ a: "a", b: "b", c: "c" }, import_string),
+        { a: "a", b: "b", c: "c" },
+        "record of strings");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: 3 }, import_number),
+        { a: 1, b: 2, c: 3 },
+        "record of numbers");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: "c" }, import_number),
+        { a: 1, b: 2 },
+        "record of numbers with a string");
+    assert.throws(() => import_record("foo", import_string), "not a record");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for .c: Not a number: "c"',
+            'JSON validation error for : Not an object: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("validate", function(assert) {
+    assert.equal(validate("test input", "foo", import_string, ""), "foo", "string");
+    assert.equal(validate("test input", 12, import_string, ""), "", "not a string");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for test input: Not a string: 12'
+        ],
+        "console errors"
+    );
+});
+
+interface Player {
+    name: string;
+    age?: number;
+}
+
+function import_Player(val: JsonValue): Player {
+    const obj = import_json_object(val);
+    const res: Player = {
+        name: import_mandatory(obj, "name", import_string),
+    };
+    import_optional(res, obj, "age", import_number);
+    return res;
+}
+
+interface Team {
+    name: string;
+    players: Player[];
+}
+
+function import_Team(val: JsonValue): Team {
+    const obj = import_json_object(val);
+    const res: Team = {
+        name: import_mandatory(obj, "name", import_string),
+        players: import_mandatory(obj, "players", v => import_array(v, import_Player))
+    };
+    return res;
+}
+
+interface Teams {
+    [name: string]: Team;
+}
+
+function import_Teams(val: JsonValue): Teams {
+    return import_record(val, import_Team);
+}
+
+QUnit.test("objects", function(assert) {
+    const checks: [JsonValue, Teams][] = [
+        [
+            {
+                MAC: {
```

Comment:
Sorry, I just have to ask: What does "MAC" mean in soccer context?

Reply:
It's supposed to be the short form of "ManCity"...